### PR TITLE
correct default random engine parameters

### LINKED ISSUE #2000: SIGSEGV with getDefaultRandomEngine()
When I try to use getDefaultRandomEngine() I get a SIGSEGV. Example code:

```C++
#include <arrayfire.h>

int main() {
    af::randomEngine engine = af::getDefaultRandomEngine();
    return 0;
}
```

And gdb log:
```
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from bad...(no debugging symbols found)...done.
(gdb) run
Starting program: /home/ralf/devel/rcpparrayfire/tmp/bad 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffd8aa3700 (LWP 11019)]
[New Thread 0x7fffd4dec700 (LWP 11020)]
[Thread 0x7fffd4dec700 (LWP 11020) exited]
[New Thread 0x7fffd4dec700 (LWP 11021)]
[New Thread 0x7fffd2501700 (LWP 11031)]
[New Thread 0x7fffd2100700 (LWP 11032)]
[New Thread 0x7fffd1cff700 (LWP 11033)]

Thread 1 "bad" received signal SIGSEGV, Segmentation fault.
0x00007fffdebadb7b in ?? () from /usr/local/lib/libafopencl.so.3
(gdb) backtrace
#0  0x00007fffdebadb7b in ?? () from /usr/local/lib/libafopencl.so.3
#1  0x00007fffdebae480 in af_retain_array () from /usr/local/lib/libafopencl.so.3
#2  0x00007ffff794e1df in ?? () from /usr/local/lib/libaf.so.3
#3  0x00007ffff796adaa in af_retain_array () from /usr/local/lib/libaf.so.3
#4  0x00007fffdee30681 in af_retain_random_engine () from /usr/local/lib/libafopencl.so.3
#5  0x00007ffff794e1df in ?? () from /usr/local/lib/libaf.so.3
#6  0x00007ffff7b21c81 in af::getDefaultRandomEngine() () from /usr/local/lib/libaf.so.3
#7  0x0000555555554965 in main ()
(gdb) 
```

I am using ArrayFire 3.5.1 (installed with the binary installer) on Debian.

### PR DESCRIPTION
**Pull request by 9prady9**
Fixes #2000

### CODE REVIEW

**Comment by pavanky** [comment]
@9prady9 you are not deleting the created handles in a destructor. Is that acceptable?

**Comment by 9prady9** [comment]
damn it, forgot about them. thanks

**Comment by 9prady9** [comment]
@pavanky please check now.

**Comment by 9prady9** [comment]
@umar456 Please check it again.

File: src/api/c/random.cpp [hunk 1]
```diff
@@ -24,24 +24,39 @@
 using namespace detail;
 using namespace common;
 
+using af::dim4;
+
+Array<uint> emptyArray()
+{
+    AF_CHECK(af_init());
```
Comment (umar456) [inline src/api/c/random.cpp]: This should only be done at public facing APIs

File: src/api/c/random.cpp [hunk 2]
```diff
@@ -24,24 +24,39 @@
...
+
+    static const Array<uint> EMPTY_ARRAY = createEmptyArray<uint>(af::dim4(0));
+
+    return EMPTY_ARRAY;
+}
+
 class RandomEngine
```
Comment (umar456) [inline src/api/c/random.cpp]: Since everything is public you should just make this a struct.

File: src/api/c/random.cpp [hunk 3]
```diff
@@ -63,14 +78,8 @@ template<typename T>
...
     if (e->type == AF_RANDOM_ENGINE_MERSENNE_GP11213) {
-        return getHandle(uniformDistribution<T>(dims,
-                    getArray<uint>(e->pos),
-                    getArray<uint>(e->sh1),
-                    getArray<uint>(e->sh2),
-                    e->mask,
-                    getArray<uint>(e->recursion_table),
-                    getArray<uint>(e->temper_table),
-                    getArray<uint>(e->state)));
+        return getHandle(uniformDistribution<T>(dims, e->pos, e->sh1, e->sh2, e->mask,
```
Comment (umar456) [inline src/api/c/random.cpp]: much better.

File: src/api/c/random.cpp [hunk 4]
```diff
@@ -118,29 +121,23 @@ af_err af_create_random_engine(af_random_engine *engineHandle, af_random_engine_
     try {
         AF_CHECK(af_init());
         validateRandomType(rtype);
+
         RandomEngine e;
```
Comment (umar456) [inline src/api/c/random.cpp]: I know you didn't add this but the way the handles are created seems inefficient. Could you look into improving it?

File: src/api/c/random.cpp [hunk 5]
```diff
@@ -154,22 +151,8 @@ af_err af_retain_random_engine(af_random_engine *outHandle, const af_random_engi
...
-
-        AF_CHECK(af_retain_array(&out.pos, engine.pos));
-        AF_CHECK(af_retain_array(&out.sh1, engine.sh1));
-        AF_CHECK(af_retain_array(&out.sh2, engine.sh2));
-        out.mask = engine.mask;
-        AF_CHECK(af_retain_array(&out.recursion_table, engine.recursion_table));
-        AF_CHECK(af_retain_array(&out.temper_table, engine.temper_table));
-        AF_CHECK(af_retain_array(&out.state, engine.state));
-
+        RandomEngine out = engine;
```
Comment (umar456) [inline src/api/c/random.cpp]: This line seems redundant.

File: src/api/c/random.cpp (hunk 1 above)
Comment (9prady9) [reply src/api/c/random.cpp]: I agree too, but without initialization OpenCL backend is causing segmentation faults as we can't create Arrays without valid context/device ids.

File: src/api/c/random.cpp (hunk 5 above)
Comment (9prady9) [reply src/api/c/random.cpp]: Fixed that.

File: src/api/c/random.cpp (hunk 4 above)
Comment (9prady9) [reply src/api/c/random.cpp]: Sure.

File: src/api/c/random.cpp (hunk 1 above)
Comment (umar456) [reply src/api/c/random.cpp]: Probably because its missing from af_get_default_random_engine

**Comment by 9prady9** [comment]
yes, that could be the case, testing it out. Also, I printed out reference counts and it seems to be working as expected. We are reusing single empty array, earlier we were creating an empty array every single time a create happened of random engine type had to change.

**Comment by 9prady9** [comment]
@umar456  I tried the following code before and after the changes in this PR.
```cpp
#include <arrayfire.h>
#include <cstdio>
#include <cstdlib>

using namespace af;

void bench(void)
{
    randomEngine re;

    re.setType(AF_RANDOM_ENGINE_THREEFRY_2X32_16);
    re.setType(AF_RANDOM_ENGINE_MERSENNE_GP11213);
    re.setType(AF_RANDOM_ENGINE_PHILOX);
    re.setType(AF_RANDOM_ENGINE_MERSENNE);
    re.setType(AF_RANDOM_ENGINE_DEFAULT);
}

int main(int argc, char *argv[])
{
    try {
        af::setDevice(0);
        printf("timeit for randomEngine modifications %g\n", timeit(bench));
    } catch (af::exception& e) {
        fprintf(stderr, "%s\n", e.what());
        throw;
    }
    return 0;
}
```

Before
```
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.00771848
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000597072
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000782411
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.0113507
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000832299
```

After
```
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000673925
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000698116
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000684257
$./build/examples/helloworld/helloworld_cuda 
timeit for randomEngine modifications 0.000665413
```

**Pull request merged**